text/0018-open-context-provider-framework.md

```diff
@@ -0,0 +1,157 @@
+          * On error, this calls `stack.reportMissingContext()` with one of the args being a cast of the `provider` from a string to
+        a `cxschema.ContextProvider`
+              * `reportMissingContext()` throws an error if the `provider` isn't from the built-in enum, and then pushes the error as
+           a `cxschema.MissingContext` onto `this._missingContext`
+          * `MissingContext` defines the `provider` as type `ContextProvider` 
+
+## Proposed changes
+
+### In CDK CLI
+
+1. Add field `readonly identifier: string` to `interface ContextProviderPlugin`
+2. Modify all existing implementations of `ContextProviderPlugin` as follows:
+    1. Define the `identifier` field with appropriate value from `cxschema.ContextProvider` enum.
+    2. Add a call to `registerContextProvider()` to the bottom of each definition file.
+        * ex: In `context-provider/vpc.ts` we add the call 
+      `registerContextProvider(VpcNetworkContextProviderPlugin.identifier, VpcNetworkContextProviderPlugin);`
+3. Modify the definition of `availableContextProviders` in `context-providers/index.ts` to be initially empty.
+4. Modify the implementation of `registerContextProvider` in `context-providers/index.ts` to throw an error if trying to re-register a name.
+5. Use the existing `PluginHost` mechanism to import/register ; extend `PluginHost` with a `registerContextProviderSource()` method.
+6. In `registerContextProviderSource()` we:
+    1. Check that the given module implements `interface ContextProviderPlugin` as best we can, and throw an error if it does not.
+    2. Check whether the plugin has already been registered. If it has, then should we throw an error here, or a warning, or... ?
+    3. Invoke `registerContextProvider()` to register the plugin
+
+### In CDK modules
+
+A saving grace here for the Construct API is that even though `enum ContextProvider` in `@aws-cdk/cloud-assembly-schema/lib/context-queries.ts`
+is a static enum that defines the names of each of the built-in `ContextProvider`s, the `interface GetContextKeyOptions` defines the `provider`
+field as being a string. So, we shouldn't require any API changes here.
+
+A challenge that I can see is that the call to `stack.reportMissingContext()` in `ContextProvider.getValue()` casts the `provider` value to a
+`cxschema.ContextProvider` (i.e. to the hard-coded enum), and then the call-chain under that requires that the `provider` be of the
+`cxschema.ContextProvider` type. For a plugable `ContextProvider` to work, all of the `providers` in this call-chain would have to have their
+type changed to `string`. I don't *think* that this will negatively impact the `CloudAssembly` -- the value of the `provider` is a string in the
+`cdk.out` json file, so changing it to a string internally should not impact that.
+
+# Drawbacks
+
+This change would alter the internal API in that some fields that are currently defined as `cxschema.ContextProvider` would need to change to
```

Comment:
Probably a better way is to add an enum value like `cxschema.ContextProvider.CUSTOM` and put the actual provider in a separate field (property). That way it's very clear what's going on.